core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/http/pipeline/stages/MakeAsyncHttpRequestStage.java

```diff
@@ -91,34 +85,57 @@ public MakeAsyncHttpRequestStage(SdkHttpResponseHandler<OutputT> responseHandler
+        //because it calls prepare() under the hood and we guarantee that we call that once per execution. It would be good
+        //to find a way to prevent multiple calls to transformResult() within a single execution to only call prepare() once.
+        final ResponseHandler handler = new ResponseHandler(responseHandler.transformResult(), errorResponseFuture);
 
-        SdkHttpRequestProvider requestProvider = context.requestProvider() == null
-                ? new SimpleRequestProvider(request, context.executionAttributes())
+        SdkHttpContentPublisher requestProvider = context.requestProvider() == null
+                ? new SimpleHttpContentPublisher(request, context.executionAttributes())
                 : context.requestProvider();
         // Set content length if it hasn't been set already.
         SdkHttpFullRequest requestWithContentLength = getRequestWithContentLength(request, requestProvider);
 
-        AbortableRunnable abortableRunnable = sdkAsyncHttpClient.prepareRequest(requestWithContentLength, SdkRequestContext
-                                                                                    .builder().build(),
-                                                                                requestProvider,
-                                                                                handler);
+        AsyncExecuteRequest executeRequest = AsyncExecuteRequest.builder()
+                .request(requestWithContentLength)
+                .requestContentPublisher(requestProvider)
+                .responseHandler(handler)
+                .build();
 
-        // Set the abortable so that the abortable request can be aborted after timeout if timeout is enabled
-        completable.abortable(abortableRunnable);
+        CompletableFuture<Void> httpClientFuture = sdkAsyncHttpClient.execute(executeRequest);
 
-        if (context.apiCallTimeoutTracker() != null && context.apiCallTimeoutTracker().isEnabled()) {
-            context.apiCallTimeoutTracker().abortable(abortableRunnable);
-        }
+        CompletableFuture<Response<OutputT>> transformFuture = handler.transformResult();
+
+        CompletableFuture<Response<OutputT>> responseFuture = new CompletableFuture<>();
+        setupAttemptTimer(responseFuture, context);
+
+        // Forward the cancellation
+        responseFuture.whenComplete((r, t) -> {
+            if (t != null) {
+                httpClientFuture.completeExceptionally(t);
+            }
+        });
+
+        // Offload the completion of the future returned from this stage onto
+        // the future completion executor
+        transformFuture.whenCompleteAsync((r, t) -> {
+            if (t == null) {
+                responseFuture.complete(r);
+            } else {
+                if (t instanceof CompletionException) {
+                    t = t.getCause();
+                }
+                responseFuture.completeExceptionally(t);
```

Comment:
This pattern is all over the place. Is there any way we can prevent its spread?

Reply:
Yep I'll add a util method in `CompletableFutureUtils`

Reply:
BTW we need this check because the type of `t` seems to depend on how the future was completed:

For a future `f`, if it's completed directly using `f.completeExceptionally(new MyException())`, then `t` is type `MyException`.

but if it's some some other future completed higher up, such as

```java
CompletableFuture<String> f0 = new CompletableFuture<>();

CompletableFuture<String> f = f0.thenApply(s ->upperCase(s));

f0.completeExceptionally(new MyException());

f.whenComplete((r, t) -> { .. });
```
then `t` will be type `CompletionException` and cause is `MyException`